gridcomps/ExtData2G/ExtDataUpdatePointer.F90

```diff
@@ -71,22 +88,72 @@ subroutine create_from_parameters(this,update_time,update_freq,update_offset,tim
          this%last_ring = this%reference_time
          this%update_freq = string_to_esmf_timeinterval(update_freq,_RC)
       end if
-      negative_offset = .false.
-      if (index(update_offset,"-") > 0) then
-         negative_offset = .true.
-         neg_index = index(update_offset,"-")
-      end if
-      if (negative_offset) then
-         this%offset=string_to_esmf_timeinterval(update_offset(neg_index+1:),_RC)
-         this%offset = -this%offset
+      i = index(update_offset,"-") + 1
+      negative_offset = i > 1
+      call parse_heartbeat_timestring(update_offset(i:), is_heartbeat=is_heartbeat, multiplier=multiplier)
+      if(is_heartbeat) then
+         this%offset = multiplier * timestep
       else
-         this%offset=string_to_esmf_timeinterval(update_offset,_RC)
+         this%offset=string_to_esmf_timeinterval(update_offset(i:),_RC)
       end if
+      if(negative_offset) this%offset = -this%offset
       _RETURN(_SUCCESS)
       _UNUSED_DUMMY(clock)
 
    end subroutine create_from_parameters
 
+   subroutine parse_heartbeat_timestring(timestring, is_heartbeat, multiplier)
+      character(len=*), intent(in) :: timestring
+      logical, intent(out) :: is_heartbeat
+      integer, intent(out) :: multiplier
+      character(len=:), allocatable :: found_string
+
+      multiplier = 1
+      call split_on(to_upper(timestring), HEARTBEAT_STRING, found_string=found_string)
+      is_heartbeat = len(found_string) > 0
+      ! For now, multiplier is simply set to 1. In the future, as needed, the before_string
+      ! and after_string arguments of split_on can be used to parse for a multiplier.
+
+   end subroutine parse_heartbeat_timestring
+
+   subroutine split_on(string, substring, found_string, before_string, after_string)
+      character(len=*), intent(in) :: string, substring
+      character(len=:), allocatable, intent(out) :: found_string
+      character(len=:), optional, allocatable, intent(out) :: before_string, after_string
+      integer :: i
+
+      i = index(to_upper(string), substring)
+      found_string = ''
+      if(i > 0) found_string = string(i:i+len(substring)-1)
+      if(present(before_string)) then
+         before_string = ''
+         if(i > 1) before_string = string(:i-1)
+      end if
+      if(present(after_string)) then
+         after_string = ''
+         if(i + len(substring) <= len(string)) after_string = string(i+len(substring):)
+      end if
+
+   end subroutine split_on
+      
+   function to_upper(s) result(u)
```

Comment:
ESMF has a built in function to upper case strings so you can just use that.

Reply:
Thanks. Done.